pkg/pattern/drain/chunk.go

```diff
@@ -94,31 +101,50 @@ func (c Chunk) ForRange(start, end, step model.Time) []logproto.PatternSample {
 	return aggregatedSamples
 }
 
-func (c *Chunks) Add(ts model.Time) {
+func (c *Chunks) Add(ts model.Time, metadata push.LabelsAdapter) {
 	t := truncateTimestamp(ts, TimeResolution)
 
 	if len(*c) == 0 {
-		*c = append(*c, newChunk(t))
+		*c = append(*c, newChunk(t, metadata))
 		return
 	}
 	last := &(*c)[len(*c)-1]
 	if last.Samples[len(last.Samples)-1].Timestamp == t {
 		last.Samples[len(last.Samples)-1].Value++
+		for _, md := range metadata {
+			last.StructuredMetadata[md.Name] = md.Value
+		}
 		return
 	}
 	if !last.spaceFor(t) {
-		*c = append(*c, newChunk(t))
+		*c = append(*c, newChunk(t, metadata))
 		return
 	}
 	last.Samples = append(last.Samples, logproto.PatternSample{
 		Timestamp: t,
 		Value:     1,
 	})
+	for _, md := range metadata {
+		last.StructuredMetadata[md.Name] = md.Value
+	}
 }
 
-func (c Chunks) Iterator(pattern string, from, through, step model.Time) iter.Iterator {
+func (c Chunks) Iterator(pattern string, from, through, step model.Time, labelFilters log.StreamPipeline) iter.Iterator {
 	iters := make([]iter.Iterator, 0, len(c))
+	var emptyLine []byte
 	for _, chunk := range c {
+		chunkMetadata := make([]labels.Label, 0, len(chunk.StructuredMetadata))
+		for k, v := range chunk.StructuredMetadata {
+			chunkMetadata = append(chunkMetadata, labels.Label{
+				Name:  k,
+				Value: v,
+			})
+		}
+		_, _, matches := labelFilters.Process(0, emptyLine, chunkMetadata...)
```

Comment:
This feels a little hacky since I'm passing empty TS and Line, but it enables re-use of the existing LogQL logic to implement the matching so I think its the best option.